Change call to smooth and remove hardcoded params for gencost_bpv4

antnguyen13 opened the pull request:
## What changes does this PR introduce?
Allow runtime parameters specific to model grid resolution for smoothing and change call to pkg/smooth that is consistent with other ecco costs.


## What is the current behaviour? 
1. Currently, bottom pressure cost that uses smoothing is relying on hardcoded numbers 300000. _d 0 and 3000, numbers that are very specific to eccov4 set up and the llc90 grid horizontal resolution.  For other model resolutions, these numbers can either be non appropriate or importantly producing Infinity/NaN smooth fields.  
2. In addition, cost_bpv4 is the only one calling smooth_basic2D while all other ecco costs (sshv4, sstv4, generic) uses smoothing calls smooth_hetero2d.  

## What is the new behaviour 
1. This PR allows the hardcoded numbers 300000. _d 0 and 3000 (corresponding to a "smooth distance scale" gencost_posproc_r and "smooth diffusion loop number" gencost_posproc_i, respectively) to be run-time.  In addition, the user MUST also set gencost_posproc='smooth' for smoothing to be applied to cost bp.
  Assuming it is gencost number 6 in data.ecco, this step (1) will require:
  gencost_posproc(?,6) = 'smooth',
  gencost_posproc_r(?,6) = 300000.0,
  gencost_posproc_i(?,6) = 3000,

2. In addition, the PR enables the user to also choose between calling the default smooth_basic2D by simply set the three parameters mentioned in (1) or calling the more general smooth_hetero2d by replacing the parameter gencost_posproc_r in (1) with a 2D distance scale stored in a file specified to parameter gencost_posproc_c.  Text has been added to the doc to show an example of how to generate this type of distance scaling file using the model grid DXC and DYC, based on my knowledge of how this file was produced for eccov4 (and subsequently for my higher res ASTE configuration).
  Assuming it is gencost number 6 in data.ecco, if the user wants to call the default smooth_basic2D, then:
  gencost_posproc(?,6) = 'smooth',
  gencost_posproc_r(?,6) = 300000.0,
  gencost_posproc_i(?,6) = 3000,
  gencost_posproc_c(?,6) = ' ',

  If want to call the more general smooth_hetero2D, then:
  gencost_posproc(?,6) = 'smooth',
  gencost_posproc_i(?,6) = 3000,
  gencost_posproc_c(?,6) = 'scaling_file_name'
 
## Does this PR introduce a breaking change? 
This can produce a change if the user got NaN or some unphysical costs before for bpv4 due to the hardcoded 300000. _d 0 and 3000, and now their more carefully chosen values can allow more physically reasonable but different costs.

## Other information:
In my ASTE configuration, I got NaN using the default hardcoded numbers.  Even if I didn't get NaN, these numbers are not necessary appropriate for my model domain.  In addition, I also checked and found only small differences between call to smooth_basic2D and smooth_hetero2D.

I also make a note that latest code made the switch from maskc(:,:,1,:,:) to maskInC in several costs, but it appeared in bpv4 there was a mixture of maskc(:,:,1,:,:) and maskInC, so likely some inconsistency in masking can exist here.  In my mod, i use maskc to be consistent.

## Suggested addition to `tag-index`
Replace hard-coded smooth-related parameters for gencost_bpv4 with run-time parameters; allow users choose between calling smooth_basic2D (using a constant distance) or smooth_hetero2D (using a 2D field to be read in).

timothyas commented on pkg/ecco/ECCO_OPTIONS.h:
```diff
@@ -62,6 +62,9 @@ C generate more text in STDOUT.0000
 C allow cost function term for sigmaR
 #undef ALLOW_SIGMAR_COST_CONTRIBUTION
 
+C debug flag from ATN
+#undef ALLOW_ECCO_ATN_DEBUG
```
I have a soft suggestion that this either be renamed to something with bpv4 in it, like `ALLOW_ECCO_BPV4_VERBOSE`, or print those debug messages when debug level is >= 3, or something like that.

antnguyen13 replied:
Thanks Tim! sure, i'll make the change.. was just something quick i put in so testers can switch to define if they want to see my print outs, but now i see they can be useful for many users.. I have been hesitant to add more debug specific, and wonder if it should also be lumped into ECCO_VERBOSE.. what do you think?

timothyas replied:
I like that idea!

timothyas commented on pkg/ecco/cost_gencost_bpv4.F:
```diff
@@ -149,6 +158,11 @@ subroutine cost_gencost_bpv4(
 c--   ============
 
+#ifdef ALLOW_ECCO_ATN_DEBUG
+      WRITE(msgBuf,'(A,1x,i10)') ' bpv4, nmonsrec: ', nmonsrec
```
It may be nice to add `kgen` here as well, if I'm understanding this code correctly.

antnguyen13 replied:
will add kgen, thanks for suggestion!

timothyas commented on pkg/ecco/cost_gencost_bpv4.F:
```diff
@@ -286,14 +311,25 @@ subroutine cost_gencost_bpv4(
+catn same syntax as well as make the call more general using runtime params.
         if ( useSMOOTH )
-     &   call smooth_basic2D(bpdifanom,maskInC,300000. _d 0,3000,myThid)
+     &   call smooth_hetero2d(bpdifanom,maskc,
+     &     gencost_posproc_c(k2_bp,kgen),
+     &     gencost_posproc_i(k2_bp,kgen),mythid)
```
I'm curious what you think about using the even more general `smooth_diff2d` routine? Since both `smooth_hetero2d` and `smooth_basic2d` have relatively small tweaks to `smooth_diff2d`.

On that note... Possibly in a future PR, I think that the options that get passed to the smooth package only be set in `data.smooth`, and all that is provided here is the integer smooth operator number. Right now smooth_hetero2d takes the `dist_file` which is being provided by `gencost_posproc_c`, but I think that this should all be dealt with in `data.smooth`. What do you think?

Either way this is an improvement from `smooth_basic2d` with the hard coded numbers, and is now consistent with `cost_gencost_sshv4`.

antnguyen13 replied:
I agree with you on somehow linking everything to what was used in data.smooth (e.g., length scale, integer smooth operator number), because I recalled having to spend some time figuring out how the offline dist_file was generated (some person-to-person knowledge from 2011 hidden in some folder... now should be documented perhaps...)

Relatedly, recently i did reroute some of these params to take just things data.smooth set.. but haven't yet found that code yet.. 

But in general, I think perhaps for this specific PR, whether we still want to "remain consistent" with other gencosts, e.g.,  sshv4, sstv4, cost_generic, in using the dist_file (however it was generated) as well as hetero2d call.. and that for a future PR soon after this, we'll then discuss to swap them all out for just smooth_diff2d?

antnguyen13 replied:
yes, @owang01 , i'd like to check with you as well since it would require you to consider using the same dist_file you have been using for cost sshv4 now for bpv4 as well.  At least it would enable you to choose how to calculate this file, and therefore document it (see my updated doc text)? instead of relying on the hardcoded numbers that I'm unsure who put it, and whether you/ecco folks are even aware that that's what you're using for bpv4 and has been very different from all other costs.

timothyas commented on pkg/ecco/cost_gencost_bpv4.F:
```diff
@@ -286,14 +311,25 @@ subroutine cost_gencost_bpv4(
+     &   call smooth_hetero2d(bpdifanom,maskc,
+     &     gencost_posproc_c(k2_bp,kgen),
+     &     gencost_posproc_i(k2_bp,kgen),mythid)
+catn     &   call smooth_basic2D(bpdifanom,maskInC,300000. _d 0,3000,myThid)
```
I suggest to delete this line rather than comment it out.

antnguyen13 replied:
i'll delete

timothyas commented on pkg/ecco/cost_gencost_bpv4.F:
```diff
@@ -286,14 +311,25 @@ subroutine cost_gencost_bpv4(
+     &   call smooth_hetero2d(bpdatanom,maskc,
+     &     gencost_posproc_c(k2_bp,kgen),
+     &     gencost_posproc_i(k2_bp,kgen),mythid)
+catn     &   call smooth_basic2D(bpdatanom,maskInC,300000. _d 0,3000,myThid)
```
Same, I think this last line can be deleted

antnguyen13 replied:
will delete

timothyas reviewed (commented):
Thanks @antnguyen13! I think this is useful fix. I have some detailed comments for you which I'm curious what you think about.

antnguyen13 commented:
@owang01 I've updated the code to enable reversion back to eccov4 set up when the dist_file (gencost_posproc_c) is missing.  So what it will take now is simply input params gencost_posproc_[r,i] = [300000. and 3000].  Could you check if eccov4 is reproducible with this PR when you have a chance?  What is required for your end will be to add gencost_posproc_[r,i] to your data.ecco instead of relying on hardcoded numbers which I have removed.

@timothyas I've resolved all your suggestions.  note, through discussion with @owang01 I've put back the ability to call smooth_basic2D so that eccov4 can be reproducible.

timothyas commented:
That sounds OK to me. After we do some cleanup with the smooth package then we can convert these code blocks into a single call to a smooth package routine, without checking for whether files exist or not. The `gencost_posproc_i` can just refer to the smooth operator number and all options can be specified in `data.smooth`. We can leave that for the future though, I guess

Thanks @antnguyen13 !

owang01 commented:
@antnguyen13 Thank you. I can reproduce V4r4 results (tested over a 3-month model integration period) using commit https://github.com/MITgcm/MITgcm/pull/693/commits/703bb752ed49ac93bd59382dfe1353de5c2974a3.

jm-c commented:
@antnguyen13 and @owang01 : I have a question regarding the replacement of the 2 hard coded number "(300000. _d 0, 3000)" in `smooth_basic2D` call: To reproduce the same results as with master branch, one would need to set `gencost_posproc_r` and `gencost_posproc_i` to 300000. and  3000.

I thinks this PR description need to be updated to mention this change.

I don't think this "bpv4" is currently used in any verification experiment, and the closest would be from `verification_other`, in  `global_oce_llc90/input_ad.ecco_v4/data.ecco`,
https://github.com/MITgcm/verification_other/blob/master/global_oce_llc90/input_ad.ecco_v4/data.ecco
with:
>  gencost_barfile(6) = 'm_bp_mon',

Could you confirm ?
Now this cost is not really used there (not reported in "ECCO configuration" summary in STDOUT), but If I want to update this "data.ecco" after this PR is merged (I was planing to do that if needed), what should I add there ? something like:
>    gencost_posproc_r( ?, 6) = 3.E+6,
>    gencost_posproc_i( ?, 6) = 3000,

@owang01 since you manage to reproduce ecco_v4 results, you might know this ?

owang01 commented:
@jm-c   In addition to the two entries (3.E+6 should be 3.E+5), there is another entry `gencost_posproc` needed. So the complete three entries are as follows:
```
gencost_posproc( ?, 6)='smooth',
gencost_posproc_r( ?, 6) = 3.E+5,
gencost_posproc_i( ?, 6) = 3000,
```

jm-c commented:
@owang01 thanks for your input.

@antnguyen13 and @owang01 I made a new PR in verification_other: https://github.com/MITgcm/verification_other/pull/60 with an update of `data.ecco` in the last commit there.
This illustrates how one would update `data.ecco` setting related to "m_bp" to get same results after this PR is merged. If this look good it could be mentionned in this PR description.

owang01 commented:
@jm-c LGTM. Thanks. 

p.s. I now realize `gencost_posproc( ?, 6)='smooth',` was already in https://github.com/MITgcm/verification_other/blob/master/global_oce_llc90/input_ad.ecco_v4/data.ecco.

antnguyen13 commented:
@jm-c and @owang01 : thank you for your detailed review and notes.  I've updated to make the description of this mod much more clear , as well as adding suggested text for the tag-index file.  Other than this, I believe there's no more changes needed on my end, except for some cosmetic comments @jm-c will add/modify.

jm-c commented:
@antnguyen13 Thanks for the updated PR description, this looks good.
@owang01 Thanks for the review, this is very useful.

jm-c commented:
@antnguyen13 I updated the comments (in last commit). Please take a look and check if this make sense (and correct for inaccuracy).

antnguyen13 commented:
Thanks @jm-c , LGTM! (and i'm going to do git pull now so i have this on my local to avoid any more potential big mess!)

jm-c commented:
From my side, I think this PR is ready to be merged. Unless someone wants to add something, will do that soon. @antnguyen13 Do you agree ?

antnguyen13 commented:
LGTM too.

jm-c commented:
@antnguyen13 Will merge this PR tomorrow. Thanks.
